sparse/core.py

```diff
@@ -912,3 +1105,33 @@ def stack(arrays, axis=0):
 
     return COO(coords, data, shape=shape, has_duplicates=has_duplicates,
                sorted=(axis == 0) and all(a.sorted for a in arrays))
+
+
+# (c) Paul Panzer
+# Taken from https://stackoverflow.com/a/47833496/774273
+# License: https://creativecommons.org/licenses/by-sa/3.0/
+def _match_coords(a, b):
+    if len(a) == 0 or len(b) == 0:
+        return np.array([], dtype=np.uint8), np.array([], dtype=np.uint8)
+    asw = np.r_[0, 1 + np.flatnonzero(a[:-1] != a[1:]), len(a)]
+    bsw = np.r_[0, 1 + np.flatnonzero(b[:-1] != b[1:]), len(b)]
+    al, bl = np.diff(asw), np.diff(bsw)
+    na = len(al)
+    asw, bsw = asw, bsw
+    abunq = np.r_[a[asw[:-1]], b[bsw[:-1]]]
+    m = np.argsort(abunq, kind='mergesort')
+    mv = abunq[m]
+    midx = np.flatnonzero(mv[:-1] == mv[1:])
+    ai, bi = m[midx], m[midx + 1] - na
+    aic = np.r_[0, np.cumsum(al[ai])]
+    a_idx = np.ones((aic[-1],), dtype=int)
```

Comment:
Do we want to specify a particular dtype here?  Do we mind this being different on different platforms?

Reply:
I haven't dived into how this function works, but in the worst case it would be `np.min_scalar_type(len(a))`, same for the `b_idx` case for down below.

Reply:
I just saw the function again. It seems we really do need positive and negative numbers, and I have no idea how big or small they need to be. I'll make them `np.int_` for now.